sdk/keyvault/internal/challenge_policy.go

```diff
@@ -0,0 +1,189 @@
+//go:build go1.16
+// +build go1.16
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+package internal
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+)
+
+const headerAuthorization = "Authorization"
+const bearerHeader = "Bearer "
+
+type KeyVaultChallengePolicy struct {
+	// mainResource is the resource to be retrieved using the tenant specified in the credential
+	mainResource *ExpiringResource
+	cred         azcore.TokenCredential
+	transport    policy.Transporter
+	scope        *string
+	tenantID     *string
+}
+
+func NewKeyVaultChallengePolicy(cred azcore.TokenCredential, t policy.Transporter) *KeyVaultChallengePolicy {
+	return &KeyVaultChallengePolicy{
+		cred:         cred,
+		transport:    t,
+		mainResource: NewExpiringResource(acquire),
+	}
+}
+
+func (k *KeyVaultChallengePolicy) Do(req *policy.Request) (*http.Response, error) {
+	as := acquiringResourceState{
+		p:   k,
+		req: req,
+	}
+
+	if k.scope == nil || k.tenantID == nil {
+		// First request, get both to get the token
+		challengeReq, err := k.getChallengeRequest(req)
+		if err != nil {
+			return nil, err
+		}
+
+		challengeResp, err := k.transport.Do(challengeReq)
```

Comment:
What's the best way to do that? Should I pass a reference to the pipeline to call all additional policies? If I call `req.Next()` I assume that will entirely run the pipeline so that could be acceptable too right?

Reply:
Not the entire pipeline--that would recurse into this policy. `req.Next()` is the right idea. We know `req` has been through every policy ahead of this one, now we want to send it through all the following policies. One possible gotcha is that `req.Next()` treats the policy list like a queue, popping the next policy off with each call, so be sure that calling `req.next()` here doesn't affect the policy list when you call it again below.

Reply:
the challenge request should pass through the entire pipeline or the original request should pass through? The original request does but if we want the challenge request to do so then I should add a copy of the clients pipeline to this struct and use that pipeline to do any modifications of this request. Does that sound correct?